pkg/infrastructure/gcp/clusterapi/clusterapi.go

```diff
@@ -46,14 +46,20 @@ func (Provider) BootstrapHasPublicIP() bool { return true }
 // created here using the gcp sdk.
 func (p Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionInput) error {
 	// Create ServiceAccounts which will be used for machines
-	projectID := in.InstallConfig.Config.Platform.GCP.ProjectID
-
-	// ServiceAccount for masters
-	// Only create ServiceAccount for masters if a shared VPC install is not being done
-	if len(in.InstallConfig.Config.Platform.GCP.NetworkProjectID) == 0 ||
-		(in.InstallConfig.Config.ControlPlane != nil &&
-			in.InstallConfig.Config.ControlPlane.Platform.GCP != nil &&
-			in.InstallConfig.Config.ControlPlane.Platform.GCP.ServiceAccount == "") {
+	platform := in.InstallConfig.Config.Platform.GCP
+	projectID := platform.ProjectID
+
+	// Only create ServiceAccounts for masters and workers if a pre-created Service Account is not defined
+	var controlPlaneMpool, computeMpool *gcptypes.MachinePool
+	if in.InstallConfig.Config.ControlPlane != nil {
+		controlPlaneMpool = in.InstallConfig.Config.ControlPlane.Platform.GCP
+	}
+
+	if gcptypes.GetConfiguredServiceAccount(platform, controlPlaneMpool) != "" {
```

Comment:
You can also use mpool "inheritance":
```suggestion
    controlPlaneMpool := &gcptypes.MachinePool{}
    controlPlaneMpool.Set(in.InstallConfig.Config.GCP.DefaultMachinePlatform)
	if in.InstallConfig.Config.ControlPlane != nil {
		controlPlaneMpool.Set(in.InstallConfig.Config.ControlPlane.Platform.GCP)
	}

	if controlPlaneMpool.ServiceAccount != "" {
```

Reply:
Thanks @r4f4, good suggestion. Updated.